File: clients/android/core-session/src/main/kotlin/de/juloc/anilingo/core/session/PlaybackSessionModels.kt
package de.juloc.anilingo.core.session

data class PlaybackSessionToken(
    val surface: String,
    val termId: String?,
    val canonical: String?,
    val reading: String?,
    val meaning: String?,
    val state: String,
)

data class PlaybackSessionSnapshot(
    val sessionId: String,
    val episodeId: String,
    val animeTitle: String,
    val episodeTitle: String,
    val positionMs: Long,
    val durationMs: Long?,
    val isPlaying: Boolean,
    val playbackRate: Double,
    val audioTrackId: String?,
    val subtitleTrackId: String?,
    val currentCueId: Long?,
    val currentCueText: String?,
    val currentCueTokens: List<PlaybackSessionToken>,
    val selectedTermId: String?,
    val revision: Long,
    val updatedAtUtc: String,
)

data class PlaybackSessionUpdate(
    val animeTitle: String,
    val episodeTitle: String,
    val positionMs: Long,
    val durationMs: Long?,
    val isPlaying: Boolean,
    val playbackRate: Double = 1.0,
    val audioTrackId: String?,
    val subtitleTrackId: String?,
    val currentCueId: Long?,
    val currentCueText: String?,
    val currentCueTokens: List<PlaybackSessionToken>,
    val selectedTermId: String?,
)

data class PlaybackSessionOwner(
    val state: PlaybackSessionSnapshot,
    val hubUrl: String,
)

data class PlaybackPairing(
    val sessionId: String,
    val code: String,
    val token: String,
    val companionUrl: String,
    val expiresAtUtc: String,
)

data class PlaybackParticipant(
    val sessionId: String,
    val accessToken: String,
    val state: PlaybackSessionSnapshot,
    val hubUrl: String,
)

data class PlaybackCommand(
    val commandId: String,
    val sessionId: String,
    val expectedRevision: Long,
    val type: String,
    val payload: Map<String, String> = emptyMap(),
    val sentAtUtc: String,
)

data class PlaybackCommandResponse(
    val acceptance: String,
    val state: PlaybackSessionSnapshot?,
    val command: PlaybackCommand?,
)

File: src/AniLingo.Web/Features/PlaybackSessions/PlaybackSessionModels.cs
namespace AniLingo.Web.Features.PlaybackSessions;

public sealed record PlaybackSessionState(
    Guid SessionId,
    Guid OwnerProfileId,
    Guid EpisodeId,
    long PositionMs,
    long? DurationMs,
    bool IsPlaying,
    double PlaybackRate,
    string? AudioTrackId,
    string? SubtitleTrackId,
    long? CurrentCueId,
    string? CurrentCueText,
    Guid? SelectedTermId,
    long Revision,
    DateTimeOffset UpdatedAtUtc);

public sealed record PlaybackSessionUpdate(
    long PositionMs,
    long? DurationMs,
    bool IsPlaying,
    double PlaybackRate,
    string? AudioTrackId,
    string? SubtitleTrackId,
    long? CurrentCueId,
    string? CurrentCueText,
    Guid? SelectedTermId);

public sealed record PlaybackPairingMaterial(
    Guid SessionId,
    string Code,
    string Token,
    DateTimeOffset ExpiresAtUtc);

public sealed record PlaybackParticipantGrant(
    Guid SessionId,
    string AccessToken,
    PlaybackSessionState State);

public sealed record PlaybackSessionCommand(
    Guid CommandId,
    Guid SessionId,
    long ExpectedRevision,
    string Type,
    IReadOnlyDictionary<string, string> Payload,
    DateTimeOffset SentAtUtc);

public enum PlaybackCommandAcceptance
{
    Accepted,
    Duplicate,
    StaleRevision,
    SessionNotFound,
    Unauthorized
}

public sealed record PlaybackCommandResult(
    PlaybackCommandAcceptance Acceptance,
    PlaybackSessionState? State,
    PlaybackSessionCommand? Command)
{
    public bool Accepted => Acceptance == PlaybackCommandAcceptance.Accepted;
}

public enum PlaybackPairingFailure
{
    None,
    InvalidOrExpired,
    RateLimited
}

public sealed record PlaybackPairingResult(
    PlaybackPairingFailure Failure,
    PlaybackParticipantGrant? Grant)
{
    public bool Success => Failure == PlaybackPairingFailure.None && Grant is not null;
}
